sdn: fix multi-master race by getting HostSubnets from event queue store

dcbw opened the pull request:
An event queue watches a single master/etcd (due to its persistent HTTP
connection), but explicit Get/List/etc calls may be directed to a different
master/etcd since they initiate a new connection.  This means the
explicit HostSubnet Get() calls in the SDN master object's addNode/deleteNode
calls might retrieve an inconsistent state.  Fix that by using the object
from th event queue's internal store.

No other parts of the SDN code explicitly call Get(), and those that
call List() do so before any of the event queues have started, so only
the HostSubnet code needs this adjustment.

Fixes: https://github.com/openshift/origin/issues/11740

dcbw commented:
@openshift/networking @danwinship

danwinship commented:
> An event queue watches a single master/etcd (due to its persistent HTTP
> connection), but explicit Get/List/etc calls may be directed to a different
> master/etcd since they initiate a new connection

Hm... so what happens when the persistent HTTP connection gets broken and it has to reconnect? Is there a chance of inconsistency at that point?

danwinship commented:
conceptually the patch looks good, but it seems messy API-wise. Could we just have RunEventQueue() pass the store to the process func instead?

dcbw commented:
> conceptually the patch looks good, but it seems messy API-wise. Could we just have RunEventQueue() pass the store to the process func instead?

@danwinship I thought about that, but the problem is that we need the HostSubnets store from the Hosts process function, so passing it to the process doesn't help because the Hosts process function would get the Hosts store, not the HostSubnets store.

Yes, we could assign the right store into the Master object on the first pass through the Hosts process function, but that's messy too since then we'd have to check for nil in a couple places since we're not guaranteed the store is populated at any particular point.

Maybe the better solution is really to just use NewEventQueueForStore() (which I removed while simplifying) and create a new store and pass that in, but that's another API change for all callers of RunEventQueue too.  Got a preference?

dcbw commented:
> Hm... so what happens when the persistent HTTP connection gets broken and it has to reconnect? Is there a chance of inconsistency at that point?

@danwinship I believe the queue makes a new connection to a (possibly different) etcd and does a Sync operation to refetch the entire list, which then shows up as add/update/delete events because we have the internal cache of the old objects.  So yeah, if it hits up an etcd with an older view of the cluster, we'll probably delete the HostSubnet for a node that we may have previously added, but then once that etcd catches up we'll add the HostSubnet back.

If etcd instances really don't agree with each other and can't stay in sync, I'm not sure how we avoid this problem.  We have to trust that something has a view of the world, and that thing is the etcd we're currently talking to.  @liggit is this completely off-base?  Do clustered etcds at least keep consistent resource versions?  How do other controllers handle this issue when they resync from a new etcd?

danwinship commented:
That sounds right. The etcds are all eventually consistent, they just don't necessarily all reflect the changes at the same time.

danwinship commented:
> @danwinship I thought about that, but the problem is that we need the HostSubnets store from the [Nodes] process function, so passing it to the process doesn't help because the [Nodes] process function would get the [Nodes] store, not the HostSubnets store.

oh, I missed that... That's a problem then because there's no guarantee that the HostSubnet store has been filled in yet by the first time the Nodes event queue looks at it.

dcbw commented:
> oh, I missed that... That's a problem then because there's no guarantee that the HostSubnet store has been filled in yet by the first time the Nodes event queue looks at it.

@danwinship to be clear, you don't mean there's a problem with the current code right?

danwinship commented:
I do mean that there's a problem with the current code; at startup, the node and hostsubnet stores get initialized in parallel by watchNodes() and watchSubnets(), so it's possible we'll get the cache.Added event for a node before the cache.Added for its HostSubnet, and so getHostSubnetByNodeName() will return an error, and addNode() will try to create the HostSubnet again and fail because it already exists and we'll log an error. (I guess nothing bad happens beyond the spurious error though.)

Also, thinking about this more, this patch doesn't fix the bug anyway: the original problem was:

t=3, node deletion event is delivered, SDN master deletes the HostSubnet
t=4, node creation event is delivered, SDN master calls HostSubnets.Get(), but talks to a different etcd master which has not yet seen the deletion at t=3, so it mistakenly thinks that the HostSubnet still exists

Using the EventQueue's store at t=4 doesn't help, because we didn't use the EventQueue's HTTP connection when we deleted the HostSubnet, so it's still possible that the two steps talk to different etcds.

dcbw commented:
> I do mean that there's a problem with the current code; at startup, the node and hostsubnet stores get initialized in parallel by watchNodes() and watchSubnets(), so it's possible we'll get the cache.Added event for a node before the cache.Added for its HostSubnet, and so getHostSubnetByNodeName() will return an error, and addNode() will try to create the HostSubnet again and fail because it already exists and we'll log an error. (I guess nothing bad happens beyond the spurious error though.)

@danwinship I don't see that...  the subnet store (which is the only one of the two that gets accessed across goroutines) is created in SubnetStartMaster() before the watchNodes/watchSubnets goroutines start...

```
	master.subnetQueue, master.subnetStore = NewSDNEventQueue()

	go utilwait.Forever(master.watchNodes, 0)
	go utilwait.Forever(master.watchSubnets, 0)
```

> Using the EventQueue's store at t=4 doesn't help, because we didn't use the EventQueue's HTTP connection when we deleted the HostSubnet, so it's still possible that the two steps talk to different etcds.

To this part, yes. I'm missing a piece, which is to remote the HostSubnet from the store when deleting it.  Then when we get t=4 creation it won't be in the store and we'll try to create it.

But the create can then fail if we're talking to an etcd with a slightly older view of the world. So I guess we need to put that into a retry loop.

danwinship commented:
> @danwinship I don't see that... the subnet store (which is the only one of the two that gets accessed across goroutines) is created in SubnetStartMaster() before the watchNodes/watchSubnets goroutines start...

But it's created *empty*. So it's possible that the watchNodes() goroutine will look at it before the watchSubnets() goroutine has finished filling it in, meaning addNode() will think the HostSubnet record doesn't exist yet when it really does.

> But the create can then fail if we're talking to an etcd with a slightly older view of the world.

That can't happen; writes force a consensus to be reached before committing.

dcbw commented:
> But it's created empty. So it's possible that the watchNodes() goroutine will look at it before the watchSubnets() goroutine has finished filling it in, meaning addNode() will think the HostSubnet record doesn't exist yet when it really does.

@danwinship even so, there's no guarantee that what's in the store at any point matches up with the current state of reality. Worst case scenario:

1. node and hostsubnet get deleted, etcds reach consensus
2. node gets created, hostsubnet gets created, only some etcds see the events
3. master gets restarted, subnet store is empty
4. watchNodes finally gets node creation event from old etcd
5. node isn't in hostsubnet store, master tries to create hostsubnet on up-to-date etcd
6. hostsubnet creation returns EEXIST; prints a warning, but things are fine
7. watchSubnets() finishes initial sync, still no hostsubnet for the new node
8. old etcd catches up, issues subnet Added event
9. watchSubnets() calls addNode/addHostSubnet
10. new hostsubnet is already in the store due to eventqueue internal logic
11. now we get an Update for the node that was previously added, but the addHostSubnet() logic is a NOP because the IP is the same, so the node gets added to nodeAddressMap.  Or if it was Deleted it's still a NOP because we don't care about nodeAddressMap for deletions.

I think the only undesirable side-effect here is the log message when the master would try to create the hostsubnet that already exists?

danwinship commented:
> there's no guarantee that what's in the store at any point matches up with the current state of reality

But if we're not even going to *try* to make it work, then why bother with the added complication of using the store at all? Just always try to create the HostSubnet when we see a new Node, and always ignore the error if its EEXIST.

(Although FWIW, the race condition I describe above seems to not actually happen in practice; for some reason, all of the HostSubnet Sync events show up before any of the Node Sync events, every time. Maybe something deep in the guts of k8s makes Node watches take longer to set up than HostSubnet watches? Anyway, adding `time.Sleep(time.Second)` to the start of watchSubnets reliably results in a bunch of `Error creating subnet 10.131.0.0/23 for node openshift-node-1: hostsubnets "openshift-node-1" already exists`, so the race condition is at least theoretically possible.)

dcbw commented:
> But if we're not even going to try to make it work, then why bother with the added complication of using the store at all?

@danwinship good question. I'm not sure it's worth attempting to make it work with complicated code given that explicit queries can always be different the event queue watch.  We might as well just try the create and if we happen to get errors.IsAlreadyExists() == true then we just try an update instead.  It should be the same info anyway.  Sound OK?

danwinship commented:
I had meant that as a rhetorical question. I think I'd want buy-in from higher-ups before switching to always-create-without-checking-cache

knobunc commented:
@smarterclayton do you have any opinion on this?

smarterclayton commented:
Two things:

1. Eventqueue needs to die - avoiding creating new things around it and gradually moving things out is good.
2. Treating "alreadyexists" as a no-op is a good thing - if the cleanup code is separate and driven by a separate loop, that means this core loop is level driven and safe.  But ultimately you can't fix a race between addition and cleanup across two types without additional info on one or both objects (so you can avoid cleaning up new objects).

Just looking through the issue, a simpler solution is usually better in these sorts of cases.  Try to break this into two loops logically - the one that guarantees one exists if the other exists, and the one dealing with cleanup.  If you handle nodes in predictable order and also handle the "someone already allocated this subnet" case you should be ok to simply attempt the create.

Creating something with a unique key is the one strong guarantee we have.  It's enough of a signal to ensure consistent controllers in some allocations (but not all)

knobunc commented:
@smarterclayton just to clarify, this is the SDN eventqueue at pkg/sdn/plugin/eventqueue.go that is built on DeltaFIFO (not the other event queue that the router uses).
